File: sol/416/416.cpp
struct Solution {
    bool canPartition(vector<int>& nums) {
        int n = nums.size(), total = reduce(nums.begin(), nums.end());
        if (total & 1){
            return false;
        }
        int target = total >> 1;
        vector<bool> memo(target + 1, false);
        memo[target] = true;
        for (int x : nums){
            for (int i = 0; i <= target - x; ++i){
                memo[i] = memo[i] || memo[i + x];
            }
        }
        return memo[0];
    }
};

File: sol/416/416.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total & 1:
            return False
        seen = set([0])
        for x in nums:
            newSeen = [val + x for val in seen]
            seen.update(newSeen)
        return (total // 2) in seen
